### ISSUE TITLE
v3 plan

### ISSUE THREAD

**Issue by jsumners** [fastify/fastify#2056]
What is the plan for releasing v3? Do we have a desired timeline?

cc: @fastify/fastify 

---

Edit from @mcollina, TODO list:

* [ ] port plugins #2217 
* [ ] preparare announcment blog post
* [ ] drop the -rc part
* [ ] change the dist-tag of all the plugins in #2217 from what is in `next` to `latest`

**Comment by delvedor** [comment]
As soon as everything in our [milestone](https://github.com/fastify/fastify/milestone/3) is closed :)

Then there are also the following pr, that needs to be merged before we can release v3.
- https://github.com/mcollina/avvio/pull/92
- https://github.com/fastify/fastify-express/pull/1
- https://github.com/fastify/middie/pull/36

**Comment by jsumners** [comment]
We need to issue at least a couple of `-pre` releases.

**Comment by Ethan-Arrowood** [comment]
TypeScript Support is in a ready state. Still work to be done, but I believe any discrepancies from this point can easily be fixed in patch or minor releases. Docs are making good progress too.

**Comment by alinnert** [comment]
@Ethan-Arrowood may I ask a question about the TypeScript support you mentioned? Does this also tackle issue #1417 in some form? (I think augmenting types is not a real solution. There are several problems with this approach.)

I'm just looking for a framework that is *fully* TypeScript compatible, which is *really* hard (i.e. no `any` type when I access things that should have a known type). I hope you can solve that somehow.

**Comment by Ethan-Arrowood** [comment]
@alinnert why do you think type augmentation isn’t a real solution? 

Decoration happens at runtime (after compile time). The types you define with TypeScript are stripped out at compile time. The only way to make decorated code type safe is to use type augmentation or type casting. I know these solutions might not be _perfect_ but this is the state of static typing a highly dynamic application.

And I can very well be wrong with my assumptions. If you have ideas on how to achieve better type safety for our decorators then please do open an issue/pull request!

**Comment by alinnert** [comment]
@Ethan-Arrowood If you augment the type `FastifyInstance` this is global. But each plugin uses its own instance with different possible decorators. Also, it can be difficult to get all properties and their types if you use 3rd party plugins. You'd need to keep this type in sync with your code and external code.

I haven't used Fastify heavily yet but I put some thought into how to solve this in general. I only have some vage ideas but, yes, I can open an issue for that later.

**Comment by jsumners** [comment]
Please take this to the linked issue or another issue about your TS concerns. I consider it off topic for this thread. And, frankly, I prefer to mute any thread about TS.

**Comment by delvedor** [comment]
Hello everyone! I wanted to share some plan for the close future about the v3 release:
We will get merged soon every [semver major](https://github.com/fastify/fastify/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3Asemver-major) pr in the `next` branch, once we do that, we will cut a 2.x branch from master and make the next branch the new master.
To recap:
`master` => `2.x`
`next` => `master`

Then we will start releasing on npm prereleases versions of v3 with the tag `next`.
Finally, we will ask to rebase all the current pull requests against master to target 2.x, and start testing all of our ecosystem against the v3 release of Fastify.

If you have some questions, please let us know!
cc @fastify/core @fastify/plugins

**Comment by MaximeBernard** [comment]
Hi! 

Do you plan on doing alpha/beta releases before that?
Do you have any planning forecasts? 

I'm currently doing some Typescript benchmarks and V2 fells a bit short with typescript integration. Should I wait for V3 or find workarounds with V2?

**Comment by mcollina** [comment]
It's almost done. We hope to issue a first alpha in the coming weeks, and then follow a series of rc.
I would say it's worth waiting for v3 wrt TypeScript, all the types has been rewritten and simplified quite a bit.

**Comment by allevo** [comment]
Are there some migration guide lines?

**Comment by Ethan-Arrowood** [comment]
@allevo there are currently no direct v2 to v3 migration guides for TS, but the doc update for v3 types includes a bunch of examples and much more detail than before so I hope that makes up for it.

**Comment by mcollina** [comment]
A blog post/guide to migrate would be helpful. I find that those are better written by users porting an app rather than authors of the breaking changes.

**Comment by Ethan-Arrowood** [comment]
TypeScript updates including docs have no been merged 🙌from the TypeScript side of things we are ready for a release candidate 🚀

**Comment by delvedor** [comment]
@fastify/fastify today between 2pm and 3pm UTC I will switch the `next` and `master` branch, please **do not** run any read/write operation during that time.

**Comment by delvedor** [comment]
Done!
Now `master` is the active branch for `v3`, while `2.x` is the active branch for `v2`.
Please rebase your prs :)

**Comment by gino8080** [comment]
how to npm install the v3 version?

**Comment by jsumners** [comment]
> how to npm install the v3 version?

Wait for an announcement that we have published to npm or do a git based install.

**Comment by Ethan-Arrowood** [comment]
Team, my work load has lightened up so please let me know if theres anything I can do to help with V3 😁

**Comment by mcollina** [comment]
@Ethan-Arrowood we should be looking for having PRs ready to go for the TS migration in most of the plugins in this org asap, as well as a review of the TS migration docs.

**Comment by delvedor** [comment]
Hey folks! Now that https://github.com/fastify/fastify/pull/2140 has been merged, we can release rc1!
I'll do it tomorrow 7 am UTC, we will also release a brief announce blog post, stay tuned and thank you for the fantastic work!

**Comment by delvedor** [comment]
Fastify v3-rc.1 is out!

Make some noise! 🎉 
https://medium.com/@fastifyjs/fastify-v3-rc1-is-here-e804fe18cb0e

**Comment by rlittolff-md** [comment]
In medium article, link to migration guide (https://github.com/fastify/fastify/blob/migration-guide-v3/docs/Migration-Guide-V3.md) give me a 404 :-/

**Comment by mcollina** [comment]
@rlittolff-md thanks! It should be fixed now.

**Comment by jsumners** [comment]
How long are we intending to run RCs?

**Comment by mcollina** [comment]
> How long are we intending to run RCs?

I would love to get all the org modules ported to v3. If we do not find any major blocker, we can ship as soon as that is done and we have an announcement blog post.

**Comment by jsumners** [comment]
> > How long are we intending to run RCs?
> 
> I would love to get all the org modules ported to v3. If we do not find any major blocker, we can ship as soon as that is done and we have an announcement blog post.

My issue is that I would like this to be used in the wild for a while before we cut a final. In particular, I would like to run some tests of my own but just do not have time right now and likely won't for a few weeks.

**Comment by mcollina** [comment]
> My issue is that I would like this to be used in the wild for a while before we cut a final. In particular, I would like to run some tests of my own but just do not have time right now and likely won't for a few weeks.

Given how folks are upgrading, it's unlikely people will migrate to v3 before it's out of the RC phase. A few volunteers will, but that's all.

**Comment by jsumners** [comment]
> Given how folks are upgrading, it's unlikely people will migrate to v3 before it's out of the RC phase. A few volunteers will, but that's all.

😢 I know. But this release is quite significant. Especially around the handling of schema validation. I'll try to get some testing done quickly, but I can't promise anything. I'm in the middle of a big project.

**Comment by mcollina** [comment]
I've updated the plan at the top. I think we are close to release, so if there are things you'd like to look into it's time.

**Comment by aggarwaldev** [comment]
I don't know if you guys have acknowledged it but nodejs just merged QUIC initial implementation with master branch.
https://github.com/nodejs/quic/issues/390

QUIC shall be essential for implementation of HTTP/3 and node is working hard on this:
https://github.com/nodejs/node/issues/23064

I request you to consider a draft implementation for QUIC and HTTP/3 since node.js implementation is experimental right now

**Comment by mcollina** [comment]
@aggarwaldev there is still no official binary with the Quic implementation and as a result it's too late to ship an implementation. Note that Quic is the _basis_ for http/3 and there is not an http/3 implementation in core. I think it's too early right now: v3 is coming out this week (today maybe).

Would you like to open a separate issue to discuss?

**Comment by aggarwaldev** [comment]
@mcollina I am so sorry. I wasn't aware about ship date of V3. 

> Quic is the basis for http/3

Yes, I do acknowledge that and that's why I wrote **shall be essential**.

> Would you like to open a separate issue to discuss?

Sure. On it.

**Comment by delvedor** [comment]
Closing this as [v3 is out](https://medium.com/@fastifyjs/fastify-v3-released-da4601d79a5)!
